bundles/org.openhab.binding.hapero/src/main/java/org/openhab/binding/hapero/internal/discovery/HaperoDiscoveryService.java

```diff
@@ -0,0 +1,121 @@
+import org.openhab.binding.hapero.internal.HaperoBindingConstants;
+import org.openhab.binding.hapero.internal.handler.HaperoBridgeHandler;
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResult;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * The {@link HaperoDiscoveryService} is responsible for discovery of new devices.
+ *
+ * @author Daniel Walter - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = HaperoDiscoveryService.class, immediate = true, configurationPid = "discovery.hapero")
+public class HaperoDiscoveryService extends AbstractDiscoveryService implements ThingHandlerService {
+
+    private static final int DISCOVER_TIMEOUT_SECONDS = 5;
+    private static final String DEVICEID = "deviceID";
+
+    private @NonNullByDefault({}) HaperoBridgeHandler bridgeHandler;
+
+    /**
+     * Constructor
+     */
+    public HaperoDiscoveryService() {
+        super(Collections.singleton(new ThingTypeUID(HaperoBindingConstants.BINDING_ID, "-")), DISCOVER_TIMEOUT_SECONDS,
+                true);
+    }
+
+    @Override
+    public void activate() {
+        super.activate(null);
+    }
+
+    @Override
+    public void deactivate() {
+        super.deactivate();
+    }
+
+    @Override
+    public @Nullable ThingHandler getThingHandler() {
+        return bridgeHandler;
+    }
+
+    @Override
+    public void setThingHandler(ThingHandler handler) {
+        if (handler instanceof HaperoBridgeHandler bh) {
+            bridgeHandler = bh;
+            bridgeHandler.setDiscoveryService(this);
+        }
+    }
+
+    @Override
+    public void startScan() {
+        // Query the current device list from the bridge
+        String[] devices = bridgeHandler.getAllDevices();
+
+        // Start a background task to check for new devices
+        scheduler.execute(() -> {
+            for (String device : devices) {
+                deviceDiscovered(device);
+            }
+        });
+    }
+
+    /**
+     * Called upon discovery of a new device.
+     * Checks if the device is supported by a Handler and creates an appropriate {@link DiscoveryResult}
+     *
+     * @param deviceId Identifier of the device that has been discovered
+     */
+    private void deviceDiscovered(String deviceId) {
+        ThingUID bridgeUID = bridgeHandler.getThing().getUID();
+        ThingUID thingUID = null;
+        String label = "";
+
+        // Check if the device is supported
+        if (deviceId.startsWith(HaperoBindingConstants.BUFFER_ID)) {
+            thingUID = new ThingUID(HaperoBindingConstants.THING_TYPE_BUFFER, bridgeUID, deviceId);
+            label = "Buffer";
```

Comment:
I18N could be supported for these labels, see for example #14577. Reference: https://github.com/openhab/openhab-addons/pull/14186#discussion_r1153048306

Reply:
According to the documentation on openhab.org it already is, because the discovery.* items are present in the hapero.properties file.

Reply:
I did it exactly how the documentation explains it. See https://www.openhab.org/docs/developer/utils/i18n.html under "Discovery".
If i try to change it to what i guessed out of the link you provided, there is no label anymore. So, either the documentation should be updated or you need to give me a more detailed explanation of how you would like it to be changed.

Reply:
I will use the same example as mentioned initially, because it exactly applies here. But too be specific:
https://github.com/openhab/openhab-addons/blob/9aab0f7d12dde8002a06aa4ef160122415a54071/bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/discovery/DanfossAirUnitDiscoveryService.java#L144-L145

and the referenced text:
https://github.com/openhab/openhab-addons/blob/9aab0f7d12dde8002a06aa4ef160122415a54071/bundles/org.openhab.binding.danfossairunit/src/main/resources/OH-INF/i18n/danfossairunit.properties#L81-L83

However, please also pay attention to the constructor:
https://github.com/openhab/openhab-addons/blob/9aab0f7d12dde8002a06aa4ef160122415a54071/bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/discovery/DanfossAirUnitDiscoveryService.java#L63-L69

You need to make sure `LocaleProvider` and `TranslationProvider` are injected.

Reply:
OK, the missing LocaleProvider and Translation Provider might be the cause this did not work. The documentation on the Hompage should definitly be updated. But then, how do i deal with multiple instances of e.g. a heating circuit ? According to the documentation, the i18n should automatically pick the right label using the binding path, thats why the labels have weird names like discovery.hapero.heatingCircuit.haperoBridge.HK1.label. The example from Danfoss above would not work if the Documentation is correct here.

Reply:
Do i need to manually parse the device ID (e.g. HK1, HK2, etc) to pick the right label "Heating Circuit 1", "Heating Circuit 2", etc. ? Thats what the method in the documentation is supposed to to for me, i thought ? My understanding was that the labels i give hardcoded in the source are just placeholders that will be replaced by i18n during runtime.

Reply:
Side question, do i need to inject the LocaleProvider and TranslationProvider in every class i want to use i18n ?

Reply:
This is how i changed it now:
`// Check if the device is supported
        if (deviceId.startsWith(HaperoBindingConstants.BUFFER_ID)) {
            thingUID = new ThingUID(HaperoBindingConstants.THING_TYPE_BUFFER, bridgeUID, deviceId);
            label = "@text/discovery.hapero.buffer.haperoBridge.PU1.label";
        } else if (deviceId.startsWith(HaperoBindingConstants.BOILER_ID)) {
            thingUID = new ThingUID(HaperoBindingConstants.THING_TYPE_BOILER, bridgeUID, deviceId);
            label = "@text/discovery.hapero.boiler.haperoBridge.WW1.label";
        } else if (deviceId.startsWith(HaperoBindingConstants.HEATING_ID)) {
            thingUID = new ThingUID(HaperoBindingConstants.THING_TYPE_HEATING, bridgeUID, deviceId);
            label = "@text/discovery.hapero.heatingCircuit.haperoBridge.HK1.label";
        } else if (deviceId.startsWith(HaperoBindingConstants.FURNACE_ID)) {
            thingUID = new ThingUID(HaperoBindingConstants.THING_TYPE_FURNACE, bridgeUID, deviceId);
            label = "@text/discovery.hapero.furnace.haperoBridge.SI.label";`

Reply:
> The documentation on the Hompage should definitly be updated. But then, how do i deal with multiple instances of e.g. a heating circuit ? According to the documentation, the i18n should automatically pick the right label using the binding path, thats why the labels have weird names like discovery.hapero.heatingCircuit.haperoBridge.HK1.label. The example from Danfoss above would not work if the Documentation is correct here.

You have a point, now re-reading the documentation this suggests another approach than I have proposed, so I can understand the confusion. I will need to dig further into the example in the documentation to verify if that works, and understand how.

Reply:
> Side question, do i need to inject the LocaleProvider and TranslationProvider in every class i want to use i18n ?

No, it's usually not needed. For example, you can use `@text/` references in `updateStatus()` without doing that.

Reply:
OK, i'll wait for your feedback with the commit on this, thanks!